jupyter/notebooks/Data Traversal.ipynb

```diff
@@ -0,0 +1,583 @@
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "1b996359-c8b3-49c2-aac4-1d13e63d337f",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "print(len(response['hits']['hits']))\n",
+    "last_result = response['hits']['hits'][20]\n",
+    "print(last_result['_id'])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "6c4c9bc7-3ac6-4f70-95cc-f410d7eebd78",
+   "metadata": {},
+   "source": [
+    "The hits does only have 20 elements. I'm confused. Is there any tool that I can use check the response JSON in a human eye frinedly way? (not the leeeeengthy data XD)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c6e78411-5f23-428c-aa4b-03b5c18d7712",
+   "metadata": {},
+   "source": [
+    "Any way, its functionally working. Time to load the process in a function!"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d3b56bf4-e7e2-4e95-addc-9f097f125b79",
+   "metadata": {},
+   "source": [
+    "## Functions and Script"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "53f930fc-2d6a-4dae-a500-2de28e4e05fa",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import os\n",
+    "from elasticsearch import Elasticsearch\n",
+    "from piffle.iiif import IIIFImageClient \n",
+    "from io import BytesIO\n",
+    "import httpx\n",
+    "from PIL import Image\n",
+    "\n",
+    "#connect to ElasticSearch\n",
+    "local_es = Elasticsearch(\n",
+    "    hosts=os.environ['LOCAL_HOST'],\n",
+    "    http_auth=(\n",
+    "        os.environ['LOCAL_USER'],\n",
+    "        os.environ['LOCAL_PASS']\n",
+    "    )\n",
+    ")\n",
+    "\n",
+    "#create a new index\n",
+    "local_es.indices.create(\n",
+    "    index='id_url_index', \n",
+    "    body={\n",
+    "        \"settings\" : {\n",
+    "            \"refresh_interval\": -1,\n",
+    "            \"number_of_replicas\": 0\n",
+    "        },\n",
+    "    }\n",
+    ")\n",
+    "index_name = 'id_url_index'\n",
+    "\n",
+    "\n",
+    "\n",
+    "def Extracting_url_ID(response): #run through each page and get the url\n",
+    "    for doc in response['hits']['hits']:\n",
+    "        ID_url_dict = {'id':doc['_id'],'url':doc['_source']['state']['derivedData']['thumbnail']['url']}\n",
+    "        yield ID_url_dict #this is convenient!\n",
+    "\n",
+    "def index_traversing(es_object, size, index_name):\n",
+    "    #run for 1 time and get a last_result\n",
+    "    response = es_object.search(\n",
+    "        index=os.environ['INDEX_NAME'],\n",
+    "        body={\n",
+    "            \"query\":{\"match_all\":{}},\n",
+    "            \"size\":size,\n",
+    "            \"from\":0,\n",
+    "            \"sort\":{\n",
+    "                \"_id\":\"asc\", #document ID. Unique(I guess)\n",
+    "            },\n",
+    "        }\n",
+    "    )\n",
+    "    #post the first 20 dictionaries\n",
+    "    for ID_url_dict in Extracting_url_ID(response):\n",
+    "        es_object.create(\n",
```

Comment:
I'm guessing you've already seen this, but it might be useful to know that you can also run `es_object.index()` ([docs here](https://elasticsearch-py.readthedocs.io/en/v7.13.3/api.html?highlight=create#elasticsearch.Elasticsearch.index)) in situations where you might want to overwrite an existing document - for experiments where you're testing stuff out repeatedly and don't need to retain every single record, this might be a more useful option